File: common/src/main/java/org/sandboxpowered/silica/resources/ZIPResourceLoader.java
package org.sandboxpowered.silica.resources;

import org.jetbrains.annotations.NotNull;
import org.sandboxpowered.api.util.Identity;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.Enumeration;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

public class ZIPResourceLoader implements ResourceLoader {
    public static Pattern pattern = Pattern.compile("(assets|data)\\/([a-z]*)\\/");
    private final File file;
    private final ZipFile zip;
    private Set<String> namespaces;

    public ZIPResourceLoader(File file) throws IOException {
        this.file = file;
        this.zip = new ZipFile(file);
    }

    @Override
    public boolean containsFile(ResourceType type, String path) {
        return false;
    }

    @Override
    public InputStream openFile(ResourceType type, String path) throws IOException {
        return null;
    }

    @Override
    public Set<String> findResources(ResourceType type, String namespace, String path, int depth, Predicate<String> filter) {
        return null;
    }

    @Override
    public Set<String> getNamespaces(ResourceType type) {
        if (namespaces == null) {
            Enumeration<? extends ZipEntry> enumeration = zip.entries();
            HashSet<String> namespaces = new HashSet<>();

            while (enumeration.hasMoreElements()) {
                ZipEntry zipEntry = enumeration.nextElement();
                String string = zipEntry.getName();
                if (string.startsWith(type.getFolder() + '/')) {
                    Matcher matcher = pattern.matcher(string);
                    if (matcher.find()) {
                        namespaces.add(matcher.group(2));
                    }
                }
            }

            this.namespaces = namespaces;
        }
        return this.namespaces;
    }

    @NotNull
    @Override
    public String getFilename(@NotNull ResourceType type, @NotNull Identity identity) {
        return null;
    }
}

File: common/src/main/kotlin/org/sandboxpowered/silica/resources/ZIPResourceLoader.kt
package org.sandboxpowered.silica.resources

import org.sandboxpowered.utilities.Identifier
import java.io.File
import java.io.InputStream
import java.util.function.Predicate
import java.util.regex.Pattern
import java.util.zip.ZipFile

class ZIPResourceLoader(override val name: String, private val file: File) : AbstractResourceLoader() {
    private val zip: ZipFile = ZipFile(file)
    private var namespaces: Set<String>? = null

    override fun containsFile(file: String): Boolean = zip.getEntry(file) != null

    override fun openFile(file: String): InputStream = when (val zipEntry = zip.getEntry(file)) {
        null -> error("")
        else -> zip.getInputStream(zipEntry)
    }

    override fun findResources(
        type: ResourceType,
        namespace: String,
        path: String,
        depth: Int,
        filter: Predicate<String>
    ): Set<Identifier> {
        TODO("Not yet implemented")
    }

    override fun getNamespaces(type: ResourceType): Set<String> {
        if (namespaces == null) {
            val enumeration = zip.entries()
            val namespaces = HashSet<String>()
            while (enumeration.hasMoreElements()) {
                val zipEntry = enumeration.nextElement()
                val string = zipEntry.name
                if (string.startsWith(type.folder + '/')) {
                    val matcher = pattern.matcher(string)
                    if (matcher.find()) {
                        namespaces.add(matcher.group(2))
                    }
                }
            }
            this.namespaces = namespaces
        }
        return namespaces!!
    }

    companion object {
        var pattern: Pattern = Pattern.compile("(assets|data)\\/([a-z]*)\\/")
    }

}